### ISSUE TITLE
Azure AD Claims with Static Web Apps and Azure Functions

### ISSUE THREAD

**Issue by johnnyreilly** [Azure/static-web-apps#988]
**Describe the bug**

Azure AD Claims are not supplied to Azure Functions when linked with Azure Static Web Apps using the "bring your own functions" / linked backend approach.

**To Reproduce**
1. Create an Azure AD application with some custom app roles; eg: 

<img width="687" alt="image" src="https://user-images.githubusercontent.com/1010525/201856786-654e181e-85af-4ef9-9feb-c099ff937778.png">

2. Create a Static Web App and a Function App 
3. Both the SWA and FA should use the Azure AD application and be [*linked*](https://blog.johnnyreilly.com/2022/10/14/bicep-static-web-apps-linked-backends)
4. Take a look at the `/.auth/me` endpoint - note the claims; they should include one of your custom app roles. eg `Microscope.Read`
5. Take a look at the claims that the Function App endpoints receive - you'll see that it is missing custom app role claims 

**Expected behavior**

**Screenshots**

**Device info (if applicable):**

N/A

**Additional context**

I'd be happy to demo this directly and share code. I'm working with Warren Joubert of Microsoft on this and understand this to be a general problem.

[Writing this problem up here, with a workaround - ideally this shouldn't be needed](https://github.com/johnnyreilly/blog.johnnyreilly.com/pull/328)

**Comment by davide-bergamini-sevenit** [comment]
A similar problem exists when linking an App Service.
I think it will be great if it could pass the AD token "X-MS-TOKEN-AAD-ID-TOKEN" like EasyAuth do, in this way we could use Microsoft.Identity.Web.

**Comment by johnnyreilly** [comment]
Agreed @davide-bergamini-sevenit  - I suspect this is a general problem and affects all linked backends.   So would expect container apps to have a similar issue.

**Comment by johnnyreilly** [comment]
I've written up my workaround in this post: https://blog.johnnyreilly.com/2022/11/17/azure-ad-claims-static-web-apps-azure-functions

Thanks @warrenandre for this assistance.

**Comment by johnnyreilly** [comment]
Is your `test-role` role a custom claims that you've configured against your Azure AD App Registration? That's what we were / are missing.

**Comment by houlgap** [comment]
Yes, I created an application role in the App registration, then assign it to the required user/group in the corresponding Enterprise Application. 

Without the getRoles function reading the claims and returning the custom role it isn't included in the userRoles array 

![image](https://user-images.githubusercontent.com/380485/206704614-b261cb2d-d3bb-4a4c-9f4b-00882c6f66c7.png)

**Comment by houlgap** [comment]
Sorry, missed a crucial detail as it coincided with how you had named your function. 
In you staticwebapp.config.json file, you need to specify the rolesSource property to point to the GetRoles function. This is called by the platform when a user logs in.
```
 "auth": {
        "rolesSource": "/api/GetRoles",
        "identityProviders": ...
    }
```
This is called

**Comment by johnnyreilly** [comment]
Oh I see! (I think)

You're not manually calling `/api/getroles` yourself, you're implementing that function and with that in place, *other" functions will now (behind the scenes) invoke this and as a consequence have the custom claims that you've configured against your Azure AD App Registration?

https://learn.microsoft.com/en-us/azure/static-web-apps/assign-roles-microsoft-graph

That's worth knowing! And also mighty peculiar!

**Comment by houlgap** [comment]
yes, sorry for the confusion!

**Comment by johnnyreilly** [comment]
I'll try and test this out - thanks for sharing!

**Comment by johnnyreilly** [comment]
It's funny, you read the docs here: https://learn.microsoft.com/en-us/azure/static-web-apps/assign-roles-microsoft-graph#verify-custom-roles and it's not obvious how the approach you're suggesting would work. However that could totally be ashortcoming of the docs - will have to suck it and see

**Comment by gbelenky** [comment]
@houlgap - I configured roles endpoint sample you provided above using "rolesSource". 
It works like a chime - I get all claims I need including App Roles defined for the user - exactly like they appear in clients /.auth/me  
However, my API endpoints don't get this information in the x-ms-client-principal header. Any idea why? Thanks!

**Comment by gbelenky** [comment]
works for me too as @houlgap suggested. Thanks, I spent so much time with @johnnyreilly blog, but was happy to find a solution which does not involve all those Graph queries. Happy New Year

**Comment by steverhall** [comment]
We had mixed results using solution from @houlgap, then realized that some times, the "typ" of the claim was just "roles", not "http://schemas.microsoft.com/ws/2008/06/identity/claims/role". Unclear as to when or why this happens, but it happens frequently with the same browser from the same user.

Changed code to: 

```
if (claim.typ == ClaimTypes.Role || claim.typ == "roles")
{
    roles.Add(claim.val);
}
```

as our work-around.

**Comment by leevi-sa** [comment]
@houlgap @steverhall Thank you for all sharing. But I found more mysterious problem.

In my PoC, the /api/getRoles is not called at all by the SWA platform.  I can call that function directly through browser with the generate function-call url.  But through the Function's Monitor, I can see there's no other calls except my manual ones.

Do you have thoughts on this issue?

I'm wondering if it's caused by that my backend function is deployed independently and linked to the SWA afterwards. Except this issue, all the interactions among AAD, SWA and AZ Function work well as excepted.

My frontend is a simple React app. And the backend is .Net/C# AZ Function.

**Comment by steverhall** [comment]
Hi,

This function gets called by SWA during the authentication process. In order for it to be called, it must be an open API (no function key or authentication required).

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by leevi-sa** [comment]
Hi @steverhall, thank you for the direction. Yes. It works now.

The mistake I made is that I did not change the default authLevel. VS uses `AuthorizationLevel.Function` as the default value.  It's working perfectly after I change it back to `AuthorizationLevel.Anonymous`.

**Comment by johschmidt42** [comment]
Thank you for providing this approach @johnnyreilly and others. I've got a few questions:
- Will the roles returned by the `rolesSource` endpoint be added to the 'userRoles' claim? I assume so but it's not clear to me from reading the docs.
- Instead of using the managed Azure Function of the SWA, can this be a suppored BringYourOwn backend? E.g. an App Service?
- Where can I read up on the specs for creating a `rolesSource`? What's the input and what's the expected output? What happens if there is an error?
- Instead of using the user's access token, why not using a service principal that has the permission to call msgraph to look up the users aad membership

**Comment by johnnyreilly** [comment]
To be honest, it's been a while since I did this and all the knowledge I have had been composed into this post:

https://johnnyreilly.com/azure-ad-claims-static-web-apps-azure-functions

**Comment by carlin-q-scott** [comment]
Am understanding correctly that the backchannel principal header only contains claims of type role from the Web App `/.auth/me` document? I wanted to use this for a multi-tenant app, so the `tenantid` claim is the most important to me, and it's missing.

**Comment by aeromac** [comment]
Just wondering if there are any updates from Microsoft on this? It would be good to do away with workarounds. If not, can Microsoft suggest a recommended approach?

**Comment by aeromac** [comment]
Can Microsoft suggest a recommended approach here? Should the frontend send the claims to the backend? or should the backend make calls to MS Graph API to get the claims?
